Bug Ensure cache_readonly is only applied on instance

fjetter opened the pull request:
The property cache was using the attribute `_cache` which occasionally is used by other classes as a class attribute. Most prominently this is true for all subclasses of `PandasExtensionDtype`, see https://github.com/pandas-dev/pandas/blob/ce082fb04a0cf597f673d35bc47afd6c2e3c3168/pandas/core/dtypes/dtypes.py#L89

This would cause the same cash to be used for the property caching as for the extension dtype. The dtype cache is a class attribute whereas the property cache is supposed to be instance only. I fixed this by using a longer name for the property caching attribute, hoping this would lead to less collisions.

Introducing a new attribute leads to failures for Cython extension types which previously were handled with an `AttributeError`. Firstly, I believe these cases should not silently pass and secondly the previous behaviour was wrong by not executing the function w/out cache but it returned the instance of the `CachedProperty` class instead. I addressed this but enabling the caching for extension classes requires defining the attribute explicitly. I went ahead and created a mixin class for this but I'm open to other suggestions as well.


- [x] closes N/A
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

fjetter commented on pandas/tests/libs/test_lib.py:
```diff
@@ -206,3 +207,38 @@ def test_get_reverse_indexer(self):
+        assert instanceA.expensive_property == "A"
+        assert instanceA.counter == 1
+
+        assert instanceB.expensive_property == "B"
```
on master this would return `A`

jbrockmendel commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -601,6 +601,7 @@ ExtensionArray
 
 - Bug in :meth:`DataFrame.where` when ``other`` is a :class:`Series` with :class:`ExtensionArray` dtype (:issue:`38729`)
 - Fixed bug where :meth:`Series.idxmax`, :meth:`Series.idxmin` and ``argmax/min`` fail when the underlying data is :class:`ExtensionArray` (:issue:`32749`, :issue:`33719`, :issue:`36566`)
+- Bug where some properties of subclasses of :class:``PandasExtensionDtype`` where improperly cached :issue:`40329`
```
where -> were

parens around the issue tag

single-ticks around PandasExtensionDtype

jbrockmendel commented:
Wouldn't it be simpler to just not instantiate _cache on the class?

fjetter commented:
> Wouldn't it be simpler to just not instantiate _cache on the class?

We don't need a cache per instance not per class.

fjetter commented:
I believe renaming the property cache is not the best idea since this is used all over the place. The actual problem is that these extension type caches https://github.com/pandas-dev/pandas/blob/7b5957f8dc0f608c328b363efcb467671ce23a35/pandas/core/dtypes/dtypes.py#L89

are class attributes but share the same name as the commonly used instance attribute and I do not know how best to distinguish them. I would propose to rename the dtypes cache but I'm not sure if this is perceived as a breaking change or not

jbrockmendel commented:
>  I would propose to rename the dtypes cache but I'm not sure if this is perceived as a breaking change or not

It would not.  Go for it.

fjetter committed:
fix cache_readonly

jbrockmendel commented on pandas/_libs/properties.pxd:
```diff
@@ -0,0 +1,5 @@
+
+cdef class PropertyCacheMixin:
```
mixing this in to other classes is going to cause many headaches, e.g. pickle and inability to inherit from other cython classes

jbrockmendel commented:
Why not go the other direction and change PandasExtensionDtype._cache to e.g. PandasExtensionDtype._instance_cache

fjetter committed:
Add whatsnew

fjetter commented:
Sorry for not responding. I was out of office for a while.

I added a runtime check to the property cache implementation which is supposed to ensure that we're not accessing any class wide state. I'm also ensuring that we're actually using the cache for ExtensionTypes and raise an appropriate exception if we do not. This led me down the serialization implementation of the `Offset` classes where most implement a custom serialization function although there is one reduce implemented which should serve all and ensures that the classes are properly instantiated. Otherwise we run into this runtime check I implemented.

@jbrockmendel

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -895,13 +895,6 @@ cdef class Tick(SingleConstructorOffset):
 
         raise ApplyTypeError(f"Unhandled type: {type(other).__name__}")
 
-    # --------------------------------------------------------------------
```
what were these hitting? or were they at all?

fjetter replied:
I honestly do not know if these were used or not. I ran into the issue that a lot of the `BaseOffset` children were not properly initialized. I ended up initializing `_cache` in every class but noticed they should all do fine using the `__reduce__` in https://github.com/pandas-dev/pandas/blob/ce34c1cee70021792b0c1b0db2c899069bbf0c1d/pandas/_libs/tslibs/offsets.pyx#L728-L737

They are all tested in https://github.com/pandas-dev/pandas/blob/ce34c1cee70021792b0c1b0db2c899069bbf0c1d/pandas/tests/io/test_pickle.py#L170-L177

with the data of https://github.com/pandas-dev/pandas/blob/ce34c1cee70021792b0c1b0db2c899069bbf0c1d/pandas/tests/io/generate_legacy_storage_files.py#L135

jbrockmendel commented on pandas/_libs/properties.pyx:
```diff
@@ -23,12 +23,29 @@ cdef class CachedProperty:
 
         # Get the cache or set a default one if needed
-        cache = getattr(obj, '_cache', None)
+        cache = instance_cache = getattr(obj, '_cache', None)
```
why is any of this necessary?

fjetter replied:
I didn't really know how to test this and implemented this runtime check. I think it's not absolutely necessary if one follows the convention properly but I believe it is very easy to make a mistake here.

jbrockmendel replied:
```

def test_no_cache_attribute_on_class():
   for cls in list_of_all_relevant_classes_somehow():
       assert not hasattr(cls, "_cache")
```

jbrockmendel commented on pandas/core/dtypes/dtypes.py:
```diff
@@ -86,7 +86,7 @@ class PandasExtensionDtype(ExtensionDtype):
     isbuiltin = 0
     isnative = 0
-    _cache: Dict[str_type, PandasExtensionDtype] = {}
+    _cache_dtypes: Dict[str_type, PandasExtensionDtype] = {}
```
the changes in this file and in pandas/tests/dtypes/test_dtypes.py make sense, the rest i think are unnecessary and in the case of cache_readonly harmful

fjetter commented:
I get that the complexity in `properties.pyx` is not necessary. I wanted to have some kind of protection that this is not happening again and this runtime check is the only way I could come up with. Code runs either way with the changes in `dtype` but I could not come up with another way of testing this behaviour. If this is fine without a dedicated test, I can revert the cython code

fjetter commented:
I reverted the changes in the cython code base. What's left is the renaming of the `ExtensionDtype._cache`

jreback commented:
@fjetter looks fine, can you merge master and ping on green
